test/e2e/network/endpointslice.go

```diff
@@ -57,20 +57,36 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
 		Description: The discovery.k8s.io API group MUST exist in the /apis discovery document.
 		The discovery.k8s.io/v1 API group/version MUST exist in the /apis/discovery.k8s.io discovery document.
 		The endpointslices resource MUST exist in the /apis/discovery.k8s.io/v1 discovery document.
-		API Server should create self referential Endpoints and EndpointSlices named "kubernetes" in the default namespace.
+		The cluster MUST have a service named "kubernetes" on the default namespace referencing the API servers.
+		The "kubernetes.default" service MUST have Endpoints and EndpointSlices pointing to each API server instance.
 	*/
-	framework.ConformanceIt("should have Endpoints and EndpointSlices pointing to API Server", func() {
+	framework.ConformanceIt("should have Endpoints and EndpointSlices pointing to the API Server Service", func() {
 		namespace := "default"
 		name := "kubernetes"
+		// verify "kubernetes.default" service exist
+		_, err := cs.CoreV1().Services(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		framework.ExpectNoError(err, "error obtaining API server \"kubernetes\" Service resource on \"default\" namespace")
+
+		// verify Endpoints for the API servers exist
 		endpoints, err := cs.CoreV1().Endpoints(namespace).Get(context.TODO(), name, metav1.GetOptions{})
-		framework.ExpectNoError(err, "error creating Endpoints resource")
+		framework.ExpectNoError(err, "error obtaining API server \"kubernetes\" Endpoint resource on \"default\" namespace")
 		if len(endpoints.Subsets) != 1 {
 			framework.Failf("Expected 1 subset in endpoints, got %d: %#v", len(endpoints.Subsets), endpoints.Subsets)
 		}
 
 		endpointSubset := endpoints.Subsets[0]
-		endpointSlice, err := cs.DiscoveryV1().EndpointSlices(namespace).Get(context.TODO(), name, metav1.GetOptions{})
-		framework.ExpectNoError(err, "error creating EndpointSlice resource")
+
+		// verify EndpointSlices for the API servers exist
+		endpointSliceList, err := cs.DiscoveryV1().EndpointSlices(namespace).List(context.TODO(), metav1.ListOptions{
+			LabelSelector: "kubernetes.io/service-name=" + name,
+		})
+		framework.ExpectNoError(err, "error obtaining API server \"kubernetes\" EndpointSlice resource on \"default\" namespace")
+		// TODO for dual-stack apiserver we should have one EndpointSlice per each IP Family
+		// Ref: https://github.com/kubernetes/enhancements/pull/2622
+		if len(endpointSliceList.Items) != 1 {
```

Comment:
I wouldn't expect this to be hard-coded to 1, shouldn't this iterate and accumulate the endpoints / ports (potentially filtering by family)

Reply:
these slices are special since only refer to the apiserver, I think that is fair to request one object per IP family to simplify client side logic, we are not going to have more 100s apiservers and ports are hardcoded as today IIRC
The problem with IP families is that the EndpointSlice object can only be IPv4 or IPv6, in the case that we want to publish different IPs for the apiserver we have to create one per each family.

Reply:
nevermind,  since we allow endpointslices reconciler outside of the cluster that is not simple to achieve, I rewrote the test to verify that the endpoint slices and the endpoints contain exactly the same IP:Port endpoints for the apiservers